auth/src/VSCodeAzureSubscriptionProvider.ts

```diff
@@ -237,30 +237,37 @@ export class VSCodeAzureSubscriptionProvider extends vscode.Disposable implement
      *
      * @returns A client, the credential used by the client, and the authentication function
      */
-    private async getSubscriptionClient(tenantId?: string, scopes?: string[]): Promise<{ client: SubscriptionClient, credential: TokenCredential, authentication: AzureAuthentication }> {
-        const armSubs = await import('@azure/arm-resources-subscriptions');
-        const session = await getSessionFromVSCode(scopes, tenantId, { createIfNone: false, silent: true });
-        if (!session) {
+    private async getSubscriptionClient(tenantId?: string): Promise<{ client: SubscriptionClient, credential: TokenCredential, authentication: AzureAuthentication }> {
+        const initialSession = await getSessionFromVSCode(undefined, tenantId, { createIfNone: false, silent: true });
+        if (!initialSession) {
             throw new NotSignedInError();
         }
 
         const credential: TokenCredential = {
-            getToken: async () => {
-                return {
-                    token: session.accessToken,
-                    expiresOnTimestamp: 0
-                };
+            getToken: async (scopes, _options) => {
+                const session = await getSessionFromVSCode(scopes, tenantId, { createIfNone: false, silent: true });
+                if (session?.accessToken) {
+                    return {
+                        token: session.accessToken,
+                        expiresOnTimestamp: 0
+                    };
+                } else {
+                    return null;
+                }
             }
         }
 
         const configuredAzureEnv = getConfiguredAzureEnv();
         const endpoint = configuredAzureEnv.resourceManagerEndpointUrl;
 
+        const armSubs = await import('@azure/arm-resources-subscriptions');
         return {
             client: new armSubs.SubscriptionClient(credential, { endpoint }),
             credential: credential,
             authentication: {
-                getSession: () => session
+                getSession: async (scopes) => {

```

Comment:
This function needs to stay non-async because of https://github.com/microsoft/vscode-azuretools/pull/1619

Reply:
OK, I'll look at 1619 and see why.

Reply:
But can you have async code in a function who's return type isn't a Thenable/Promise?

Reply:
> I am wondering what will happen if I make the getSession function synchronous but let it return a Promise.

Tested and it makes the deploy fail just like https://github.com/microsoft/vscode-azurefunctions/issues/3878 with an "Number of entries expected in End Of Central Directory does not correspond to number of entries in Central Directory." error.

Reply:
A fix for the deploy issue has been merged. The getSession function no longer needs to be synchronous.